src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -289,36 +292,115 @@ float LightTree::evaluate_node_pdf(
     return pdf;
 }
 
+Vector3d LightTree::emitting_triangle_centroid(const size_t triangle_index) const
+{
+    const EmittingTriangle& triangle = m_emitting_triangles[triangle_index];
+    return (triangle.m_v0 + triangle.m_v1 + triangle.m_v2) * (1.0 / 3.0);
+}
+
+namespace
+{
+    // [1] Section 2.2.
+    float sub_hemispherical_light_source_contribution(
+        const float     cos_omega,
+        const float     cos_sigma)
+    {   
+        assert(cos_omega >= -1.0 && cos_omega <= 1.0);
+        assert(cos_sigma >= 0.0 && cos_sigma <= 1.0);
```

Comment:
Called `cos_sigma` here but `cos_theta` at the call site. Which is best? Can we use the same name? Maybe not, just asking.